rasa/shared/core/events.py

```diff
@@ -631,6 +632,24 @@ def as_dict(self) -> Dict[Text, Any]:
         d.update({USE_TEXT_FOR_FEATURIZATION: self.use_text_for_featurization})
         return d
 
+    def apply_to(self, tracker: "DialogueStateTracker") -> None:
+        """Applies event to current conversation state.
+
+        Args:
+            tracker: The current conversation state.
+        """
+        if tracker.latest_action_name != ACTION_LISTEN_NAME:
+            # featurization belong only to the last user message
+            # a user message is always followed by action listen
+            return
+
+        # update previous user message's featurization based on this event
+        if tracker.latest_message.use_text_for_featurization is None:
+            # only update it if it was not defined
```

Comment:
I think this kind of comments is not particularly. helpful as they just repeat the code. The question (and I'm actually having it 😆 ): Why do we need to safeguard against this and when could this actually happen that we try to set it when it was already set previously?

Reply:
every time we apply the events from tracker to tracker, which we do multiple times

Reply:
Mhm, and applying it twice is bad because of performance or why?

Reply:
I don't know 😅

Reply:
actually, can we guarantee that `UserUttered` events will be persisted with all the updates? Then I can remove `Define` events from applied events

Reply:
I removed safeguard, I have don't know whether it is possible that we would redefine featurization

Reply:
>  Then I can remove Define events from applied events

That's actually the thing I'm trying to get you to fix in [here](https://github.com/RasaHQ/rasa/pull/7503#discussion_r540236892)